numpy/lib/index_tricks.pyi

```diff
@@ -0,0 +1,179 @@
+@overload
+def ix_(*args: _NestedSequence[bytes]) -> Tuple[_ArrayND[bytes_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[bool]) -> Tuple[_ArrayND[bool_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[int]) -> Tuple[_ArrayND[int_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[float]) -> Tuple[_ArrayND[float_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[complex]) -> Tuple[_ArrayND[complex_], ...]: ...
+@overload
+def ix_(*args: _RecursiveSequence) -> Tuple[_ArrayND[Any], ...]: ...
+
+class nd_grid(Generic[_BoolType]):
+    sparse: _BoolType
+    def __init__(self, sparse: _BoolType = ...) -> None: ...
+    @overload
+    def __getitem__(
+        self: nd_grid[Literal[False]],
+        key: Union[slice, Sequence[slice]],
+    ) -> _ArrayND[Any]: ...
+    @overload
+    def __getitem__(
+        self: nd_grid[Literal[True]],
+        key: Union[slice, Sequence[slice]],
+    ) -> List[_ArrayND[Any]]: ...
+
+class MGridClass(nd_grid[Literal[False]]):
+    def __init__(self) -> None: ...
+
+mgrid: MGridClass
+
+class OGridClass(nd_grid[Literal[True]]):
+    def __init__(self) -> None: ...
+
+ogrid: OGridClass
+
+class AxisConcatenator:
+    axis: int
+    matrix: bool
+    ndmin: int
+    trans1d: int
+    def __init__(
+        self,
+        axis: int = ...,
+        matrix: bool = ...,
+        ndmin: int = ...,
+        trans1d: int = ...,
+    ) -> None: ...
+    @staticmethod
+    @overload
+    def concatenate(  # type: ignore[misc]
+        *a: ArrayLike, axis: SupportsIndex = ..., out: None = ...
+    ) -> _ArrayND[Any]: ...
+    @staticmethod
+    @overload
+    def concatenate(
+        *a: ArrayLike, axis: SupportsIndex = ..., out: _ArrayType = ...
+    ) -> _ArrayType: ...
+    @staticmethod
+    def makemat(
+        data: ArrayLike, dtype: DTypeLike = ..., copy: bool = ...
+    ) -> _Matrix: ...
+
+    # TODO: Sort out this `__getitem__` method
+    def __getitem__(self, key: Any) -> Any: ...
+
+class RClass(AxisConcatenator):
+    axis: Literal[0]
+    matrix: Literal[False]
+    ndmin: Literal[1]
+    trans1d: Literal[-1]
+    def __init__(self) -> None: ...
+
+r_: RClass
+
+class CClass(AxisConcatenator):
+    axis: Literal[-1]
+    matrix: Literal[False]
+    ndmin: Literal[2]
+    trans1d: Literal[0]
+    def __init__(self) -> None: ...
+
+c_: CClass
+
+class IndexExpression(Generic[_BoolType]):
+    maketuple: _BoolType
+    def __init__(self, maketuple: _BoolType) -> None: ...
+    @overload
+    def __getitem__(  # type: ignore[misc]
+        self: IndexExpression[Literal[True]], item: slice
+    ) -> Tuple[slice]: ...
+    @overload
+    def __getitem__(self, item: _SliceOrTuple) -> _SliceOrTuple: ...
+
+index_exp: IndexExpression[Literal[True]]
+s_: IndexExpression[Literal[False]]
+
+def fill_diagonal(a: ndarray[Any, Any], val: Any, wrap: bool = ...) -> None: ...
+def diag_indices(n: int, ndim: int = ...) -> Tuple[_ArrayND[int_], ...]: ...
```

Comment:
Both `diag_indices` and `diag_indices_from` return tuples of `int_` arrays.
Shouldn't they be returning `intp` arrays considering the values are supposed to represent indices?

Reply:
Seems they return whatever arange returns. Arange probably returns whatever `n` is? So I guess it probably will go to 64bit if the value is too large.  But I tend to agree, in the spirit of not changing the output dtype `intp` would make more sense.

Reply:
> In the spirit of not changing the output dtype `intp` would make more sense.

If there are no objects then I'd like to change `arange(...)` to `arange(..., dtype=np.intp)`.
In practice this should only affect Windows and 32-bit platforms anyway, as `int_` already
maps to `int64` on all (most?) other platforms.

Reply:
I would like to try changing the default integer to `intp` entirely :), while pretty simple, it may be a big step. Is it very valuable here? It feels a bit like we have this problem all over the place and the only reason this one jumps at you is that `_indices` is in the function name.  Although, maybe that is a great reason!

Reply:
> It feels a bit like we have this problem all over the place and the only reason this one jumps at you is that `_indices` is in the function name.

Yeah, that's basically my motivation here. Functions that return indices generally return them as `intp` arrays, this one didn't so it stood out to me.

> I would like to try changing the default integer to `intp` entirely :)

+1 to that suggestion!

Reply:
I had a PR once (with some tiny holes possible). Its not _that_ bad, it "only" affects 64bit windows, and any library should already be written in a non-insane way and not really notice the change much.

My problem is that we would have to some risk analysis probably. Its not like we can _warn_ about it... It doesn't make sense if every `np.arange` call or `np.array([3])` call spits out a warning.

It is something I would rather try with a "major release", but in cetero censeo: I think we should just plan one and hopefully tick of one or two other things.  But, I don't have the energy/motivation right now to attempt that without more encouragement.

---

About this, I am fine with going for it here (although maybe not `dtype=np.intp` which might force cast a float input).  But it needs to go through the mailing list, since it is technically an incompatible change without a deprecation period!

Reply:
> But it needs to go through the mailing list, since it is technically an incompatible change without a deprecation period!

Ah, that's something I hadn't actually, since I suppose that you don't _technically_ have to pass an integer `diag_indices`. In any case, this seemingly small change sounds like something that would benefit from a dedicated PR after all.